keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,845 @@
+scenario where users are downloading binaries from untrustworthy sources to use
+in their cluster.  However, this is not a departure from existing security
+posture of the exec credential feature present in k8s.io/client-go today.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Proxy Proposal
+
+The proxy exec plugin is a long-running proxy process which is executed by the
+client on first use.  When executed, the exec proxy plugin will bind and listen
+on a socket, and will a output URL on stdout, over which the client would proxy
+all API server bound requests.  The proxy binary location is defined in the
+`ExecConfig` section of a kubeconfig `user`.
+
+The proxy plugin process will be executed from client-go the first time it is
+needed, on transport setup.  If the client fails to connect to the socket
+because no one is listening, client-go will reattempt to exec the plugin.
+Client-go will support both plugins binaries which are themselves the proxy
+process, and additionally plugin binaries which complete, but are responsible
+for starting a proxy process and returning a URL to the proxy.
+
+The client will send its request to the proxy plugin process, and the proxy
+will forward the request to the API server, rather than returning the modified
+request to the client to be sent (see alternatives considered).  The downside
+of this approach is that the proxy must duplicate the some of the transport
+functionality of client packages, like respecting existing proxy configuration,
+but allows for more flexibility for proxy plugin developers.
+
+### Authentication to Proxy
+
+The client will authenticate to its proxy using mTLS.  On exec, the proxy will
+pass back client key and certificate authority data for the client to use for
+authentication when establishing a connection.
+
+An alternative method for authentication to the proxy would be file permission
+based.  The GID owner of the client process could passed to the proxy over stdin,
+which the proxy should give read/write permissions on the socket to.  However,
+this could be less portable if we need to use a non UDS communication mechanism
+for windows, or if we support TCP/IP in the future.
+
+### Caching
+
+The connection from the client to the proxy can be cached in the client based
+on the cluster details, the existing HTTP_PROXY configuration, the exec-proxy
+URL, and the certificates used for mTLS.
+
+### Proxy Shutdown
+
+In the case where the client dies unexpectedly, due to a crash or SIGKILL, the
+proxy should also die.  To handle this, the proxy will attempt to take a file
+lock that the parent holds and quit if it ever successfully takes the lock.
```

Comment:
Where is this in the API?

Reply:
We can also have platform specific implementation of this and use locks in the generic implementation. On linux, you can use PR_SET_PDEATHSIG:

https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/syscall/exec_linux.go;l=48-52;drc=126c22a09824a7b52c019ed9a1d198b4e7781676

Which simplifies the implementation of exec plugins that only need to work on linux. On windows, you can use job objects.

Reply:
If we need lock files, they have to be present in the API.  Also, we will provide a reference implementation so the bulk of implementers should not have to care.

Reply:
Added `LockFilePath` to the API.